server/app/survey.go

```diff
@@ -50,3 +64,197 @@ func (a *UserSurveyApp) StopSurvey(surveyID string) error {
 
 	return nil
 }
+
+func (a *UserSurveyApp) ShouldSendSurvey(userID string, survey *model.Survey) (bool, error) {
+	key := utils.KeyUserSendSurveyLock(userID)
+	locked, err := a.TryLock(key, time.Now().UTC())
+	if err != nil {
+		return false, errors.Wrap(err, "ShouldSendSurvey: failed to acquire lock for checking if survey needs to be sent to user or not")
+	}
+
+	defer func() {
+		_ = a.Unlock(key)
+	}()
+
+	if !locked {
+		return false, nil
+	}
+
+	if survey.Status != model.SurveyStatusInProgress {
+		return false, errors.New("ShouldSendSurvey: a survey can only be sent against an in progress survey")
+	}
+
+	alreadySent, err := a.getSurveySentToUser(userID, survey.ID)
+	if err != nil {
+		return false, errors.Wrap(err, "ShouldSendSurvey: failed to check if survey is already sent to the user")
+	}
+
+	if alreadySent {
+		return false, nil
+	}
+
+	inExcludedTeam, err := a.userInExcludedTeams(userID, survey)
+	if err != nil {
+		return false, errors.Wrap(err, "ShouldSendSurvey: failed to check is in filtered teams or not")
+	}
+
+	return !inExcludedTeam, nil
+}
+
+func (a *UserSurveyApp) getSurveySentToUser(userID, surveyID string) (bool, error) {
+	data, appErr := a.api.KVGet(utils.KeyUserSurveySentStatus(userID, surveyID))
+	if appErr != nil {
+		a.api.LogError("GetSurveySentToUser: Failed to get user survey sent status key from KV store", "userID", userID, "surveyID", surveyID, "error", appErr.Error())
+		return false, errors.Wrap(appErr, "GetSurveySentToUser: Failed to get user survey sent status key from KV store")
+	}
+
+	if data == nil {
+		return false, nil
+	}
+
+	return string(data) == surveySentValue, nil
+}
+
+func (a *UserSurveyApp) setSurveySentToUser(userID, surveyID string) error {
+	appErr := a.api.KVSet(utils.KeyUserSurveySentStatus(userID, surveyID), []byte(surveySentValue))
+	if appErr != nil {
+		a.api.LogError("setSurveySentToUser: Failed to set user survey sent status in KV store", "userID", userID, "surveyID", surveyID, "error", appErr.Error())
+		return errors.Wrap(appErr, "setSurveySentToUser: Failed to set user survey sent status in KV store")
+	}
+
+	return nil
+}
+
+func (a *UserSurveyApp) SendSurvey(userID string, survey *model.Survey) error {
+	if err := a.ensureSurveyBot(); err != nil {
```

Comment:
Could we do this when the plugin is initialized instead of waiting until the survey is sent?

Reply:
Done.